### PR TITLE
[Perf] Bypass detokenizer for prefill-only requests

### PR DESCRIPTION
**Pull request by narutolhy**
## Motivation

Prefill-only requests (`max_new_tokens=0`) are common in embedding, scoring, and cache-warming workloads. Currently these requests still go through the full detokenizer round-trip even though they produce zero output tokens. This adds unnecessary ZMQ hops, detokenization overhead, and a GPU→CPU sync for dummy token IDs.

### Data Flow

```
Before (all requests):

  Scheduler ──BatchTokenIDOutput──► Detokenizer ──BatchStrOutput──► Tokenizer Manager
                 (ZMQ #1)           detokenize("")    (ZMQ #2)
                                    ↑ wasted work

After (prefill-only, max_new_tokens=0):

  Scheduler ──BatchStrOutput──► Tokenizer Manager
                (ZMQ #1)
              skip detokenizer ✓
```

### GPU→CPU Transfer

```
Before:  torch.zeros(device="cuda")  ──.to("cpu")──►  .tolist()
                                        ↑ GPU sync

After:   torch.zeros() [CPU]  ──►  [0]*N   (no sync)
```

### Changes

1. **CPU-side dummy token IDs** (`tp_worker.py`): Create prefill-only dummy `next_token_ids` directly on CPU (drop `device=input_ids.device`), avoiding a GPU→CPU transfer + sync.
2. **Skip redundant `.to("cpu")` / `.tolist()`** (`utils.py`, `scheduler_output_processor_mixin.py`): Guard the CPU transfer and `tolist()` calls with `is_cpu` / `is_prefill_only` checks so we don't pay for a no-op move.
3. **Bypass detokenizer for finished prefill-only batches** (`scheduler_output_processor_mixin.py`): When every request in a batch has `max_new_tokens=0` and is already finished, send a `BatchStrOutput` (with empty `output_strs`) directly to the tokenizer manager, saving one full ZMQ round-trip through the detokenizer.

## Accuracy Tests

No change to model outputs — prefill-only requests produce no tokens; the bypass only applies when `max_new_tokens=0` and the request is already finished.

## Speed Tests and Profiling

**Setup:** H100, Qwen-0.6B, batch=16, seq_len=256, prefill-only (`max_new_tokens=0`)

| Configuration | Avg Latency | Throughput | Delta |
|---|---|---|---|
| Baseline (`main`, `disable_overlap`) | 15.42 ms | 266K tok/s | — |
| + deferred release + bypass detokenizer | 11.79 ms | 323K tok/s | **-2.76 ms (−19%)** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
```

```markdown
## Motivation

Prefill-only requests (`max_new_tokens=0`) are common in embedding, scoring, and cache-warming workloads. Currently these requests still go through the full detokenizer round-trip even though they produce zero output tokens. This adds unnecessary ZMQ hops, detokenization overhead, and a GPU→CPU sync for dummy token IDs.

## Modifications

### Data Flow

```
Before (all requests):

  Scheduler ──BatchTokenIDOutput──► Detokenizer ──BatchStrOutput──► Tokenizer Manager
                 (ZMQ #1)           detokenize("")    (ZMQ #2)
                                    ↑ wasted work

After (prefill-only, max_new_tokens=0):

  Scheduler ──BatchStrOutput──► Tokenizer Manager
                (ZMQ #1)
              skip detokenizer ✓
```

### GPU→CPU Transfer

```
Before:  torch.zeros(device="cuda")  ──.to("cpu")──►  .tolist()
                                        ↑ GPU sync

After:   torch.zeros() [CPU]  ──►  [0]*N   (no sync)
```

### Changes

1. **CPU-side dummy token IDs** (`tp_worker.py`): Create prefill-only dummy `next_token_ids` directly on CPU (drop `device=input_ids.device`), avoiding a GPU→CPU transfer + sync.
2. **Skip redundant `.to("cpu")` / `.tolist()`** (`utils.py`, `scheduler_output_processor_mixin.py`): Guard the CPU transfer and `tolist()` calls with `is_cpu` / `is_prefill_only` checks so we don't pay for a no-op move.
3. **Bypass detokenizer for finished prefill-only batches** (`scheduler_output_processor_mixin.py`): When every request in a batch has `max_new_tokens=0` and is already finished, send a `BatchStrOutput` (with empty `output_strs`) directly to the tokenizer manager, saving one full ZMQ round-trip through the detokenizer.

## Accuracy Tests

No change to model outputs — prefill-only requests produce no tokens; the bypass only applies when `max_new_tokens=0` and the request is already finished.

## Speed Tests and Profiling

**Setup:** H100, Qwen-0.6B, batch=16, seq_len=256, prefill-only (`max_new_tokens=0`)

| Configuration | Avg Latency | Throughput | Delta |
|---|---|---|---|
| Baseline (`main`, `disable_overlap`) | 15.42 ms | 266K tok/s | — |
| + deferred release + bypass detokenizer | 11.79 ms | 323K tok/s | **-2.76 ms (−19%)** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

**Comment by sundar24295s** [comment]
@narutolhy Can you help validate if this works with Multi-Item scoring? (Sample server start commands [here](https://github.com/sgl-project/sglang/pull/22427)) . Except we changed the `--multi-item-scoring-delimeter` server arg to `--enable-mis` instead.
